src/devices/regency_fan.c

```diff
@@ -0,0 +1,230 @@
+/** @file
+    Decoder for Regency fan remotes
+
+    Copyright (C) 2020-2022 David E. Tiller
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+ */
+
+/**
+ The device uses OOK_PULSE_PPM encoding.
+ The packet starts with 576 uS start pulse.
+ - 0 is defined as a 375 uS gap followed by a 970 uS pulse.
+ - 1 is defined as a 880 uS gap followed by a 450 uS pulse.
+
+ Transmissions consist of the start bit followed by bursts of 20 bits.
+ These packets ar repeated up to 11 times.
+
+ As written, the PPM code always interpets a narrow gap as a 1 and a
+ long gap as a 0, however the actual data over the air is inverted,
+ i.e. a short gap is a 0 and a long gap is a 1. In addition, the data
+ is 5 nibbles long and is represented in Little-Endian format. In the
+ code I invert the bits and also reflect the bytes. Reflection introduces
+ an additional nibble at bit offsets 16-19, so the data is expressed a 3
+ complete bytes.
+
+ The examples below are _after_ inversion and reflection (MSB's are on
+ the left).
+
+ Packet layout
+ Bit number
+ 0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16 17 18 19 20 21 22 23
+  CHANNEL  |  COMMAND  |            VALUE       | 0  0  0  0| 4 bit checksum
+
+ CHANNEL is determined by the bit switches in the battery compartment. All
+ switches in the 'off' position result in a channel of 15, implying that the
+ switches pull the address lines down when in the on position.
+
+ COMMAND is one of the following:
+
+ CMD_STOP        0x01
+        value: (0xc0, unused).
+
+ CMD_FAN_SPEED   0x02
+        value: 0x01-0x07. On my remote, the speeds are shown as 8 - value.
+
+ CMD_LIGHT_INT   0x04
+        value: 0x00-0xc3. The value is the intensity percentage.
+               0x00 id off, 0xc3 is 99% (full).
+
+ CMD_LIGHT_DELAY 0x05
+        value: 0x00 is 'off', 0x01 is on.
+
+ CMD_FAN_DIR     0x06
+        value: 0x07 is one way, 0x83 is the other.
+
+ The CHECKSUM is calculated by adding the nibbles of the first two bytes
+ and ANDing the result with 0x0f.
+
+ */
+
+#include <stdlib.h>
+#include "decoder.h"
+
+#define NUM_BITS        21
```

Comment:
expand all defines into the code.

Reply:
What is your rationale for not using #define constants? They give an awful lot of context in the code as to what the constant _means_, not just it's value.

int value = MAX_VALUE;

is a lot clearer than

int value = 37;

Reply:
what's `MAX_VALUE` then? which data does it apply to? does it change? ;)
`int value = 37; // maximum expected row length, shorter for "learn" packet`

Explain bits, nibbles, bytes in the comment text above (already looks good), that's much better than `BYTE_START`(is it a start/sync byte? or counted from where?).
The decoder code then can't avoid bit manipulation and numbers, embrace that ;) and if needed comment better than a #define could, e.g. ` case 1: // stop command`

but TL;DR, it's easier to reason about and maintain 200+ decoders without a preprocessor involved.

Reply:
In the end it's your code, but I disagree about #defines and using explicit width types. As I said there are several hundred #defines in the existing device code, not to mention the rtl_433 code itself. You explicitly use uint8_t in your function definitions, so it's not like I'm adding something that's otherwise unused.